### DISCUSSION TITLE
Framework vs SDK for displaying authenticated user info? When should I use PB Framework, and when I should use JavaScript?

### DISCUSSION THREAD

**Discussion by VictorioBerra** [pocketbase/pocketbase#4438] [Q&A]
Breaking this off into its own discussion since my last one was getting long.

Today, I am displaying the username like so:

```golang
// From cookie (set with OnRecordAuthRequest)
authRecord, _ := c.Get(apis.ContextAuthRecordKey).(*models.Record)
html, err := registry.LoadFiles(
	"views/layout.html",
	"views/"+staticPageOptions.PageName+".html",
).Render(map[string]any{
	"authRecord": authRecord,
})
```

```html
{{ with .authRecord }}
<li class="nav-item dropdown ms-auto">
  <a class="nav-link dropdown-toggle" href="#" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    {{ .Username }}
  </a>
</li>
{{ else }}
```

Now I am wondering why not call `new PocketBase('...').authStore.model.username` and display it with AlpineJS?

I guess im confusing myself on when I should use the PB Framework, and when I should use JavaScript?

**Comment by dschissler** [comment]
Why aren't you checking the error?

```golang
authRecord, _ := c.Get(apis.ContextAuthRecordKey).(*models.Record)
```

Did you do auth using the SDK?  If not, then the `new PocketBase('...').authStore.model.username` won't have any value because it's getting it from the LocalStorage that the SDK set in Javascript.  So it's only going to be available if you did the auth the conventional way.

If you're going to feed auth records into a Golang template then I would try to create a safe version of it first.  In the other thread you can see the function called `safeUser` that copies only certain specific fields into the `SafeUser` struct that is meant to be passed into templates.

**Reply by VictorioBerra** [reply]
@dschissler thanks for your reply. I will check the error.

Yes I am starting to do all auth with the JS SDK, and I set the auth cookie using `OnRecordAuthRequest()` like we discussed.

I will use the "safeUser" you mentioned [here](https://github.com/pocketbase/pocketbase/discussions/4409#discussioncomment-8584538).

Mainly I am just wondering, is it cleaner, or faster, or smarter or whatever to hide/show buttons, and show the username in the navbar on the server-side using the Go Template OR use the JS SDK to get the user info on page load, and simply use Alpine to hide/show buttons and display login info.

Hide/Show:

![image](https://github.com/pocketbase/pocketbase/assets/2934507/66d0fb09-67f0-4114-a603-3ceecce13c3d)

![image](https://github.com/pocketbase/pocketbase/assets/2934507/d41714a3-e355-4968-8ea6-43d2bffbaba9)

**Comment by VictorioBerra** [comment]
@dschissler `authRecord, _ := c.Get(apis.ContextAuthRecordKey).(*models.Record)` why bother "error" checking this?

`_` represents a bool indicating if the type assertion was successful or not. And, from my understanding it will also be false if `authRecord` is nil.

A nil `authRecord` simply means the user is not logged in. There is no situation where I would store a non-Record type in ContextAuthRecordKey.

**Reply by VictorioBerra** [reply]
Usage in the template would be like so:

```html
          {{ with .authRecord }}
          <li class="nav-item dropdown ms-auto">
            <a class="nav-link dropdown-toggle" href="#" role="button" data-bs-toggle="dropdown" aria-expanded="false">
              {{ .Username }}
            </a>
            <ul class="dropdown-menu dropdown-menu-end">
              <li><a class="dropdown-item" href="/logout">Logout</a></li>
            </ul>
          </li>
          {{ else }}
          <li class="nav-item ms-auto">
            <a class="nav-link" href="/login">Login</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="/register">Register</a>
          </li>
          {{ end }}
```

**Comment by dschissler** [comment]
> `_` represents a bool indicating if the type assertion was successful or not. And, from my understanding it will also be false if `authRecord` is nil.
> 
> A nil `authRecord` simply means the user is not logged in. There is no situation where I would store a non-Record type in ContextAuthRecordKey.

Ok you're right.

I suggest using the new(ish) API helper method.

`"github.com/pocketbase/pocketbase/apis"`

```golang
apis.RequestInfo(c).AuthRecord
```

**Comment by VictorioBerra** [comment]
@dschissler Can we just chat on Discord for like 15 minutes? lol I think it would cut down on the long disjointed discussions here.

**Reply by dschissler** [reply]
@VictorioBerra I can probably find time to chat in about 10 days.  Now I'm only allowing myself for small distractions.
